doc/monads/functors.lean

```diff
@@ -0,0 +1,215 @@
+example, let's supposed you create a new type describing the measurements of a home
+or apartment:
+
+-/
+structure LivingSpace (α: Type) where
+  totalSize : α
+  numBedrooms : Nat
+  masterBedroomSize : α
+  livingRoomSize : α
+  kitchenSize : α
+  deriving Repr, BEq
+/-!
+
+Now you can construct a `LivingSpace` in square feet using floating point values:
+-/
+abbrev SquareFeet := Float
+
+def mySpace : LivingSpace SquareFeet :=
+  { totalSize := 1800, numBedrooms := 4, masterBedroomSize := 500,
+    livingRoomSize := 900, kitchenSize := 400 }
+/-!
+
+Now, suppose you want anyone to be able to map a `LivingSpace` from one type of measurement unit to
+another.  Then you would provide a `Functor` instance as follows:
+
+-/
+def LivingSpace.map (f : α → β) (s : LivingSpace α) : LivingSpace β :=
+  { totalSize := f s.totalSize,
+    numBedrooms := s.numBedrooms,
+    masterBedroomSize := f s.masterBedroomSize,
+    livingRoomSize := f s.livingRoomSize,
+    kitchenSize := f s.kitchenSize }
+
+instance : Functor LivingSpace where
+  map := LivingSpace.map
+/-!
+
+Notice this functor instance takes `LivingSpace` and not the fully qualified type `LivingSpace SquareFeet`.
+Notice below that `LivingSpace` is a function from Type to Type.  For example, if you give it type `SquareFeet`
+it gives you back the fully qualified type `LivingSpace SquareFeet`.
+
+-/
+#check LivingSpace -- Type → Type
+/-!
+
+So the `instance : Functor` then is operating on the more abstract, or generic `LivingSpace` saying
+for the whole family of types `LivingSpace α` you can map to `LivingSpace β` using the generic
+`LivingSpace.map` map function by simply providing a function that does the more primitive mapping
+from `(f : α → β)`.  So `LivingSpace.map` is a sort of function applicator.
+
+Notice that `LivingSpace.map` applies a function `f` to convert the units of all the LivingSpace
+fields, except for `numBedrooms` which is a count (and therefore is not a measurement that needs
+converting).
+
+So now you can define a simple conversion function, let's say you want square meters instead:
+
+-/
+abbrev SquareMeters := Float
+def squareFeetToMeters (ft : SquareFeet ) : SquareMeters := (ft / 10.7639104)
+/-!
+
+and now bringing it all together you can use the simple function `squareFeetToMeters` to map
+`mySpace` to square meters:
+
+-/
+#eval squareFeetToMeters <$> mySpace
+/-
+{ totalSize := 167.225472,
+  numBedrooms := 4,
+  masterBedroomSize := 46.451520,
+  livingRoomSize := 83.612736,
+  kitchenSize := 37.161216 }
+  -/
+/-!
+
+Wow, this is pretty powerful.  By providing a functor instance on `LivingSpace` with an
+implementation of the `map` function it is now super easy for anyone to come along and
+transform the units of a `LivingSpace` using very simple functions like `squareFeetToMeters`. Notice
+that squareFeetToMeters knows nothing about `LivingSpace`.
+
+You can implement the `Functor` pattern in other languages, people have even done it in C++, but it
```

Comment:
This sentence is a bit out of place. Could you please remove it?

Reply:
Done.